gax/src/main/java/com/google/api/gax/rpc/AbstractResponseObserver.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2018, Google LLC All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.InternalApi;
+import com.google.common.base.Preconditions;
+
+/**
+ * Base implementation of {@link ResponseObserver} that performs state sanity checks.
+ *
+ * <p>This is public only for technical reasons, for advanced usage.
+ */
+@InternalApi
+public abstract class AbstractResponseObserver<V> implements ResponseObserver<V> {
```

Comment:
It seems like we could solve the problem in a more flexible way by having a concrete response observer that does checks and then forwards calls to an inner response observer that it contains, instead of having the response observers inheriting from an abstract class. (The pattern I suggest is how things generally work in GAX already.)

Reply:
I agree with you that composition is more flexible than inheritance and I'm ok making the change. However, I wanted to get your opinion from a different angle first.

The main motivation for this change was to help prevent bugs like #454, where the upstream misbehaved and it was hard to narrow down which link created the issue. By having the target ResponseObserver implement the state checks (current impl), we are saying that ResponseObservers are responsible for defending themselves.  By inserting a state checking wrapper (your approach), then we are saying that the possibly misbehaving upstream is responsible for sanitizing it's output by wrapping the target with a sanitizer. 

The downside to having the upstream sanitize its own output is that if a new misbehaving transformer is inserted into a chain that doesn't sanitize its output, the errors will show up in the next link. On the upside, it keeps each link a lot simpler.

I don't have a strong preference, so I leave it up to you. If you decide to go with the change, how do you feel about `StateCheckingResponseForwarder` as the name for the concrete class?

Reply:
Even in your code here, nothing forces upstream to extend `AbstractResponseObserver`; upstream can still directly implement the interface. In all cases, we need to depend on people copying convention when they make a new response observer. I don't know that having upstream extend `AbstractResponseObserver` really means that they are sanitizing their own output; they would be using gax's code to do the sanitization. Likewise, if upstream wraps their observer with a  `StateCheckingResponseForwarder` and that's what they pass to gax, it's still gax's code that is doing the sanitization. 

Now here's where they really diverge: a client might want to wrap the `StateCheckingResponseForwarder` with even more checks. That kind of customization is not possible with the inheritance approach, but is trivially additive in the compositional approach. 

As for the name: do you dislike `StateCheckingResponseObserver`?

Reply:
After discussing offline, we decided to keep the AbstractResponseObserver approach. The wrapping approach had issues with finding the responsible party that would do the wrapping.